src/slice.c

```diff
@@ -245,25 +245,49 @@ static SEXP vec_slice_base(enum vctrs_type type, SEXP x, SEXP index) {
   }
 }
 
-static SEXP slice_names(SEXP names, SEXP index) {
-  if (names == R_NilValue) {
-    return names;
+// Replace any `NA` name caused by `NA` index with the empty
+// string. It's ok mutate the names vector since it is freshly
+// created (and the empty string is persistently protected anyway).
```

Comment:
For these cases, should we assert that the input is a fresh vector using `NO_REFERENCES()`?